[[ Bug 16474 ]] Make odd numbers of dashes work properly

runrevmark opened the pull request:
Skia does not accept odd numbers of dashes in its path effects and
causes an assertion failure in debug mode in that case.

This has been fixed by ensuring that odd numbers of dashes set a
repeated pattern. i.e. 1,2,3 => 1,2,3,1,2,3.

runrevmark committed:
[[ Bug 16474 ]] Make odd numbers of dashes work properly.

Skia does not accept odd numbers of dashes in its path effects and
causes an assertion failure in debug mode in that case.

This has been fixed by ensuring that odd numbers of dashes set a
repeated pattern. i.e. 1,2,3 => 1,2,3,1,2,3.

peter-b commented on libgraphics/src/utils.cpp:
```diff
@@ -292,10 +297,10 @@ bool MCGDashesToSkDashPathEffect(MCGDashesRef self, SkDashPathEffect*& r_path_ef
+			t_dashes[i] = MCGFloatToSkScalar(self -> lengths[i % self -> count]);
 	
-		t_dash_effect = new SkDashPathEffect(t_dashes, self -> count, MCGFloatToSkScalar(self -> phase));
+		t_dash_effect = new SkDashPathEffect(t_dashes, (int)t_dash_count, MCGFloatToSkScalar(self -> phase));
```
Since the argument you pass to `SkDashPathEffect()` eventually needs to be an `int`, I suggest using `int t_dash_count` with the appropriate overflow checks to avoid an overflowable cast (although if someone's using a dash pattern that long I'm not sure we should worry too much about whether it overflows or not).

``` c++
int t_dash_count;
if (t_success)
{
    if (self->count % 2 > 0)
    {
        /* Skia doesn't accept odd-length dash patterns, so repeat any odd-length pattern twice */
        t_dash_count = 2 * self->count;
        t_success = (self->count <= INT_MAX / 2); /* overflow check */
    }
    else
    {
        t_dash_count = self->count;
        t_success = (self->count <= INT_MAX);
    }
}

/* ... */

if (t_success)
{
    for (int i = 0; i < t_dash_count; ++i)
        /* ... etc ... */
}
```

peter-b replied:
Eh, it probably doesn't matter all that much, since the chances of anyone using a dash pattern longer than `INT_MAX` is infinitesimal.

peter-b commented:
@livecode-vulcan review ok ae0f6f1

livecode-vulcan commented:
💙 review by @peter-b ok ae0f6f1228c4b610093b4ac91c735e97ee3acfc4

runrevmark committed:
[[ Bug 16474 ]] Fix incorrect memory allocation.

The allocation for the dash array was using self -> count rather
than t_dash_count causing a memory overflow bug if the number of
dashes was odd.

peter-b commented:
@livecode-vulcan review ok 3d3b32c

livecode-vulcan commented:
💙 review by @peter-b ok 3d3b32c7b43b625037619ed87b3aad6b7131536c

livecode-vulcan commented:
😎 test success 3d3b32c7b43b625037619ed87b3aad6b7131536c
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1282)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1293)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1278)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1288)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1299)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1298)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1303)
